src/main/java/ca/uhn/fhir/jpa/starter/interceptor/CustomAuthorizationInterceptor.java

```diff
@@ -0,0 +1,178 @@
+		return new RuleBuilder()
+			.allowAll()
+			.build();
+	}
+
+	private List<IAuthRule> unauthorizedRule() {
+		// By default, deny everything except the metadata request
+		return new RuleBuilder()
+			.allow().metadata().andThen()
+			.denyAll()
+			.build();
+	}
+
+	private List<IAuthRule> authorizeOAuth(RequestDetails theRequest) throws AuthenticationException {
+		logger.info("Authorizing via OAuth2");
+		String token = OAuth2Helper.getToken(theRequest);
+		try {
+			DecodedJWT jwt = JWT.decode(token);
+			String jwksUrl = config.getOauth().getJwks_url();
+			OAuth2Helper.verify(jwt, jwksUrl);
+
+			List<String> clientRoles = OAuth2Helper.getClientRoles(jwt, getOAuthClientId());
+			if (clientRoles.isEmpty()) {
+				logger.warn("Authorization failure - token doesn't have any client roles");
+				return unauthorizedRule();
+			}
+
+			// The only difference between the admin role and the user role is that the admin role
+			// allows DELETE requests. It still needs to enforce a patient claim, if one exists.
+			if (theRequest.getRequestType().equals(RequestTypeEnum.DELETE)
+					&& !clientRoles.contains(getOAuthAdminRole())) {
+				logger.warn("Authorization failure - token doesn't have the admin role required for delete");
+				return unauthorizedRule();
+			}
+
+			if (clientRoles.contains(getOAuthAdminRole()) || clientRoles.contains(getOAuthUserRole())) {
+				String patientId = OAuth2Helper.getClaimAsString(jwt, "patient");
+				if (Strings.isNullOrEmpty(patientId)) {
+					logger.debug("No patient claim specified in authorization token");
+					return authorizedRule();
+				} else {
+					logger.debug("Patient claim specified in in authorization token; will use patient compartment rules");
+					return authorizedInPatientCompartmentRule(theRequest, patientId);
+				}
+			}
+
+			logger.warn("Authorization failure - token doesn't have the required client roles");
+			return unauthorizedRule();
+		} catch (RuntimeException|GeneralSecurityException e) {
+			logger.warn("Authentication failure - unable to decode or verify token: {}", e.getMessage());
+			throw new AuthenticationException("Invalid authorization header", e);
+		}
+	}
+
+	private List<IAuthRule> authorizedInPatientCompartmentRule(RequestDetails theRequestDetails, String patientId) {
+		if (OAuth2Helper.canBeInPatientCompartment(theRequestDetails.getResourceName())) {
+			IdType patientIdType = new IdType("Patient", patientId);
+			return new RuleBuilder()
+				.allow().read().allResources().inCompartment("Patient", patientIdType).andThen()
+				.allow().patch().allRequests().andThen()
+				.allow().write().allResources().inCompartment("Patient", patientIdType).andThen()
+				.allow().transaction().withAnyOperation().andApplyNormalRules().andThen()
+				.denyAll()
+				.build();
+		}
+		return authorizedRule();
+	}
+
+	private boolean isOAuthEnabled() {
+		return config.getOauth().getEnabled();
+	}
+
+	private String getOAuthClientId() {
```

Comment:
Minor comment: aren;t we trying to use camelCase everywhere? Or are these fields not under our control?

Reply:
Which fields are you referring to? The line you commented on has the function declaration using camelCase. If you are referring to functions below like `getClient_id`, they are part of `AppProperties.java` and are "bound" to configuration in the `application.yaml` file. If you look at the code in `AppProperties.java`, most of the properties were defined this way. I'm just trying to be consistent with existing code.